### PR TITLE
feat: API:  support typed errors over RPC

### LINKED ISSUE #9041: Make batch message more clever
### Checklist

- [X] This is **not** a new feature or an enhancement to the Filecoin protocol. If it is,  please open an [FIP issue](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0001.md).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** brainstorming ideas. If you have an idea you'd like to discuss, please open a new discussion on [the lotus forum](https://github.com/filecoin-project/lotus/discussions/categories/ideas) and select the category as `Ideas`.
- [X] I **have** a specific, actionable, and well motivated improvement to propose.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [ ] lotus miner/market - storage deal
- [ ] lotus miner/market - retrieval deal
- [ ] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [X] Other

### Improvement Suggestion

as stated in PRs like https://github.com/filecoin-project/lotus/pull/8986

batch messages may hit block limit - lotus should be clever enough know when to chunk them and how.

### PR DESCRIPTION
**Pull request by arajasek**
## Related Issues

This is work towards, but does not finish #9041 

## Proposed Changes

- Updates go-jsonrpc.
- Introduces API errors, with only one case right now -- OutOfGas.
- Demo test.

## Additional Info

- [x] Needs https://github.com/filecoin-project/go-jsonrpc/pull/36 & tag
- [ ] The chief limitation here is that _wrapped_ errors do not get picked up by the RPC code, which relies on reflection. This is a non-trivial issue, although it is not a blocker. I'd like to discuss this with @magik6k, but it doesn't preclude landing this PR.

## Checklist

Before you mark the PR ready for review, please make sure that:
- [x] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

**Comment by magik6k** [comment]
> The chief limitation here is that wrapped errors do not get picked up by the RPC code, which relies on reflection.

We could make go-jsonrpc recursively [errors.Unwrap](https://pkg.go.dev/errors#Unwrap) the error until it hits a registered error (with some depth limit just in case), then have it return a "wrapped" version of that error:
* What the error code would be in that case is a good question, we can go one of a few routes:
  * Pre-register the 'wrapped' error type, which would marshal into the `{errorMsg, subErrCode, subErr}` - probably closest how errors are wrapped in Go, but weird in non-Go apps - you can't tell the error type just by looking at the error code in the jsonrpc response
  * Register wrapped error types for each `wrapable` error type - you can tell the error just by looking at the error code, but now there are 2 error codes per error type.
  * Mix of the above - when registering an error specify if it's wrapable - if it is, always marshal it as wrapped json, if it is not, keep current behavior.

File: cmd/lotus/daemon.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ import (
 	"github.com/filecoin-project/go-paramfetch"
 
 	"github.com/filecoin-project/lotus/api"
+	lapi "github.com/filecoin-project/lotus/api"
```
Comment (magik6k) [inline cmd/lotus/daemon.go:34]: duplicate import

Comment (arajasek) [reply cmd/lotus/daemon.go]: fixed

**Comment by magik6k** [comment]
Tests seem to be broken because of this really robust error check: https://github.com/filecoin-project/lotus/blob/master/itests/kit/blockminer.go#L194 (which we can now make actually robust with this PR)

File: lib/retry/retry_test.go [hunk 2]
```diff
@@ -7,19 +7,21 @@ import (
...
 	errorsToRetry := []error{&jsonrpc.RPCConnectionError{}}
-	require.False(t, ErrorIsIn(xerrors.Errorf("random error"), errorsToRetry))
+	require.False(t, api.ErrorIsIn(xerrors.Errorf("random error"), errorsToRetry))
 }
 
 func TestRetryWrappedErrorIsInTrue(t *testing.T) {
 	errorsToRetry := []error{&jsonrpc.RPCConnectionError{}}
-	require.True(t, ErrorIsIn(xerrors.Errorf("wrapped: %w", &jsonrpc.RPCConnectionError{}), errorsToRetry))
+	require.True(t, api.ErrorIsIn(xerrors.Errorf("wrapped: %w", &jsonrpc.RPCConnectionError{}), errorsToRetry))
 }
```
Comment (magik6k) [inline lib/retry/retry_test.go:27]: Technically those tests should also be moved to `api/`

File: itests/kit/blockminer.go [hunk 3]
```diff
@@ -191,7 +192,7 @@ func (bm *BlockMiner) MineBlocksMustPost(ctx context.Context, blocktime time.Dur
 			reportSuccessFn := func(success bool, epoch abi.ChainEpoch, err error) {
 				// if api shuts down before mining, we may get an error which we should probably just ignore
 				// (fixing it will require rewriting most of the mining loop)
-				if err != nil && !strings.Contains(err.Error(), "websocket connection closed") {
+				if err != nil && !strings.Contains(err.Error(), "websocket connection closed") && !api.ErrorIsIn(err, []error{&jsonrpc.RPCConnectionError{}}) {
```
Comment (magik6k) [inline itests/kit/blockminer.go:195]: supernit:
```suggestion
				if err != nil && !strings.Contains(err.Error(), "websocket connection closed") && !api.ErrorIsIn(err, []error{new(jsonrpc.RPCConnectionError)}) {
```

**Review by magik6k — APPROVED** [review]

**Comment by dirkmc** [comment]
Another use case where this would be helpful:
https://github.com/filecoin-project/boost/blob/a58ea5a92a4237a018f9e7dc700936a5cebb0688/cmd/booster-bitswap/remoteblockstore/remoteblockstore.go#L97-L102

**Comment by hunjixin** [comment]
sorry for these questions after the code merge

Recently venus intended to submit some of venus changes and bug fixes for rpc to the upstream, but when pushing it, it found many conflicts with TypeError

In PR, I think the current implementation of TypeError is a bit complicated.

I'm not sure if users really need the MarshalJson/UnMarshalJson functionality, but I'm looking at lotus/boost usage and it's enough to determine the type of error, and in venus requirements it's enough to determine the type of error.
in addition, I have the impression that adding meta fields may cause errors in the jsonrpc library in rust, which does not allow additional fields. meta is not a in standard jsonrpc protocol.

ErrorCode is a good idea, but maybe more better to define ErrorCode as true error type like below code

```go
type ErrorCode int

func (e ErrorCode) Error() string {
    return fmt.Sprintf("error code %d", e)
}
```

for rpc server , user fmt.Error to return specific error
```go
if res.MsgRct.ExitCode == exitcode.SysErrOutOfGas {
        return -1, rpc.ErrorCode(exitcode.SysErrOutOfGas)
}
```

or  convert all exitcode to rpc error

```go
if res.MsgRct.ExitCode != exitcode.Ok {
        return -1, xerrors.Errorf("message execution failed: exit %w, reason: %s", rpc.ErrorCode(res.MsgRct.ExitCode), res.Error)
}
```

for rpc client, use this code to check error type
```go
    require.True(t, xerrors.Is(err, rpc.EOutOfGas))
```

If  write it this way, not need to add the Register function in Rpc, and the Rpc implementation can be simpler without adding the meta field in Respnse.

**Pull request merged**